Src/VsVimShared/ICommandTarget.cs

```diff
@@ -29,7 +29,7 @@ internal interface ICommandTarget
     {
         CommandStatus QueryStatus(EditCommand editCommand);
 
-        bool Exec(EditCommand editCommand, out Action action);
+        bool Exec(EditCommand editCommand, out Func<Func<int>, int> wrapper);

```

Comment:
One of the intents of this design over IOleCommandTarget was essentially to take the concept of next out of the implementation. Essentially no need to call into the next implementation. Why do we need to add this back here?

Reply:
The `GoToDefinition` code needs to run an action *before* the command (to collect the list of text views that have pre-existing selections) and to run an action *after* the command (to clear any new selections). This can be solved by either:

1) Splitting the current action into two actions: a before hook and an after hook
2) Changing the API to a allow single wrapper that accepts an opaque function to call the target

I chose the second method because:

1) The before and after hooks for `GoToDefinition` have to communicate state between each other (namely the list of text views with selections)
2) It's a relatively easy general mechanism in the functional-style with considerable flexibility

The major advantages of the new approach is that the wrapper can:

1) Bypass the command entirely
2) Use the success or failure of the command to affect the hooked actions

Only the latter feature is currently being used.

It didn't occur to me that the before hook and after hook could communicate state by capturing a variable in a closure but I think that will work.